File: frontend/src/api/migrate.ts
import client from './client'

export interface ContainerInfo {
  id: string
  name: string
  image: string
  status: string
  state: string
  env: Record<string, string>
  ports: { host_port: number; container_port: number; protocol?: string }[]
  labels: Record<string, string>
  mounts: { source: string; target: string; read_only?: boolean }[]
}

export interface ParamValue {
  name: string
  value: unknown
}

export interface MigrationCandidate {
  container: ContainerInfo
  matched_service: string
  services: string[]
  extracted_params: ParamValue[]
}

export function analyzeMigrations(): Promise<{ data: MigrationCandidate[] }> {
  return client.get('/migrate/analyze')
}

export interface GenerateAppRequest {
  container_id: string
  service_name: string
}

export interface GenerateAppResult {
  service_name: string
  file_path: string
}

export function generateApp(req: GenerateAppRequest): Promise<{ data: GenerateAppResult }> {
  return client.post('/migrate/generate', req)
}

export interface AdoptRequest {
  container_id: string
  service_name: string
  repo_name?: string
  version?: string
  params?: ParamValue[]
  rebuild?: boolean
}

export interface AdoptResult {
  container_id: string
  service_name: string
  endpoint: string
  labels: Record<string, string>
}

export function adoptContainer(req: AdoptRequest): Promise<{ data: AdoptResult }> {
  return client.post('/migrate/adopt', req)
}

File: contracts/migrate.go
package contracts

type MigrateService interface {
	Analyze(epName string) ([]*MigrationCandidate, error)
	Generate(req *GenerateAppRequest) (*GenerateAppResult, error)
	Adopt(req *AdoptRequest) (*AdoptResult, error)
	PreviewAdopt(req *AdoptPreviewRequest) (*AdoptPreviewResult, error)
}

type MigrationCandidate struct {
	Container       *ContainerInfo `json:"container"`
	MatchedService  string         `json:"matched_service"`
	Services        []string       `json:"services"`
	ExtractedParams []*ParamValue  `json:"extracted_params"`
}

type GenerateAppRequest struct {
	ContainerID string `json:"container_id"`
	ServiceName string `json:"service_name"`
}

type GenerateAppResult struct {
	ServiceName string `json:"service_name"`
	FilePath    string `json:"file_path"`
}

// AdoptRequest is used to rebuild an existing container as a managed service.
// The original container is stopped and removed, then recreated with the same
// configuration plus managed labels. If ServiceName matches a known service
// template and Params are provided, the template install pipeline is used.
type AdoptRequest struct {
	ContainerID string        `json:"container_id"`
	ServiceName string        `json:"service_name"`
	RepoName    string        `json:"repo_name,omitempty"`
	Version     string        `json:"version,omitempty"`
	Params      []*ParamValue `json:"params,omitempty"`
}

type AdoptResult struct {
	ContainerID string            `json:"container_id"`
	ServiceName string            `json:"service_name"`
	RepoName    string            `json:"repo_name,omitempty"`
	Endpoint    string            `json:"endpoint"`
	Labels      map[string]string `json:"labels"`
}

// AdoptPreviewRequest is used to preview what an adoption would create.
type AdoptPreviewRequest struct {
	ContainerID string        `json:"container_id"`
	ServiceName string        `json:"service_name"`
	Params      []*ParamValue `json:"params,omitempty"`
}

// AdoptPreviewResult contains the preview of what would be created by Adopt.
type AdoptPreviewResult struct {
	RunParams      *ContainerRunParams `json:"run_params"`
	DockerRunCmd   string              `json:"docker_run_cmd"`
	ServiceName    string              `json:"service_name"`
	OriginalLabels map[string]string   `json:"original_labels"`
}
